updater/src/updater.rs

```diff
@@ -0,0 +1,348 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::sync::{Arc, Weak};
+use std::{fs, env};
+use std::io::Write;
+use std::path::{PathBuf};
+//use util::misc::platform;
+use ipc_common_types::{VersionInfo, ReleaseTrack};
+use util::{Address, H160, H256, FixedHash, Mutex, Bytes};
+use ethcore::client::{BlockId, BlockChainClient, ChainNotify};
+use hash_fetch::{self as fetch, HashFetch};
+use operations::Operations;
+use service::{Service};
+use types::all::{ReleaseInfo, OperationsInfo, CapState};
+
+/// Filter for releases.
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub enum UpdateFilter {
+	/// All releases following the same track.
+	All,
+	/// As with `All`, but only those which are known to be critical. 
+	Critical,
+	/// None.
+	None,
+}
+
+/// The policy for auto-updating.
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub struct UpdatePolicy {
+	/// Download potential updates.
+	pub enable_downloading: bool,
+	/// Disable client if we know we're incapable of syncing.
+	pub require_consensus: bool,
+	/// Which of those downloaded should be automatically installed.
+	pub filter: UpdateFilter,
+}
+
+impl Default for UpdatePolicy {
+	fn default() -> Self {
+		UpdatePolicy {
+			enable_downloading: false,
+			require_consensus: true,
+			filter: UpdateFilter::None,
+		}
+	}
+}
+
+#[derive(Debug, Default)]
+struct UpdaterState {
+	latest: Option<OperationsInfo>,
+
+	fetching: Option<ReleaseInfo>,
+	ready: Option<ReleaseInfo>,
+	installed: Option<ReleaseInfo>,
+
+	capability: CapState,
+}
+
+/// Service for checking for updates and determining whether we can achieve consensus.
+pub struct Updater {
+	// Useful environmental stuff.
+	update_policy: UpdatePolicy,
+	weak_self: Mutex<Weak<Updater>>,
+	client: Weak<BlockChainClient>,
+	fetcher: Mutex<Option<fetch::Client>>,
+	operations: Mutex<Option<Operations>>,
+	exit_handler: Mutex<Option<Box<Fn() + 'static + Send>>>,
+
+	// Our version info (static)
+	this: VersionInfo,
+
+	// All the other info - this changes so leave it behind a Mutex.
+	state: Mutex<UpdaterState>,
+}
+
+const CLIENT_ID: &'static str = "parity";
+
+// TODO!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! REMOVE!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+fn platform() -> String {
+	"test".to_owned()
+}
+
+impl Updater {
+	pub fn new(client: Weak<BlockChainClient>, update_policy: UpdatePolicy) -> Arc<Self> {
+		let mut u = Updater {
+			update_policy: update_policy,
+			weak_self: Mutex::new(Default::default()),
+			client: client.clone(),
+			fetcher: Mutex::new(None),
+			operations: Mutex::new(None),
+			exit_handler: Mutex::new(None),
+			this: VersionInfo::this(),
+			state: Mutex::new(Default::default()),
+		};
+
+		// TODO!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! REMOVE!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+		if u.this.track == ReleaseTrack::Unknown {
```

Comment:
yeah they'll be removed prior to merge (there's a TODO for that "Remove testing-cradle code.").